traits/trait_list_object.py

```diff
@@ -53,460 +88,750 @@ def __init__(self, index=0, removed=None, added=None):
+        then fires the event.
+
+        Parameters
+        ----------
+        trait_list : list
+            The list
+        index : int or slice
+            Index or slice that was modified
+        removed : value or list of values
+            Value or list of values that were removed
+        added : value or list of values
+            Value or list of values that were added
+
+        """
+        is_trait_none = self.trait is None
+        is_name_items_none = self.name_items is None
+        if not hasattr(self, "trait") or is_trait_none or is_name_items_none:
             return
 
-        removed = self.copy()
-        list.clear(self)
+        object = self.object()
+        if object is None:
+            return
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed)
-            )
+        # bug-for-bug conversion of parameters to TraitListEvent
+        if isinstance(index, slice):
+            if index.step in {1, None}:
+                index = min(index.start, index.stop)
+            else:
+                if added and not isinstance(added, list):
+                    added = [added]
+                if not isinstance(removed, list):
+                    removed = [removed]
+        else:
+            if removed is Undefined:
+                removed = []
+            elif not isinstance(removed, list):
+                removed = [removed]
+            if added is Undefined:
+                added = []
+            elif not isinstance(added, list):
+                added = [added]
+        event = TraitListEvent(index, removed, added)
+        items_event = self.trait.items_event()
+        object.trait_items_event(self.name_items, event, items_event)
 
-    def len_error(self, len):
-        """ Utility method that raises an error if length is incorrect. """
-        raise TraitError(
-            "The '%s' trait of %s instance must be %s, "
-            "but you attempted to change its length to %d element%s."
-            % (
-                self.name,
-                class_of(self.object()),
-                self.trait.full_info(self.object(), self.name, Undefined),
-                len,
-                "s"[len == 1:],
-            )
+    def __deepcopy__(self, memo):
+        """ Perform a deepcopy operation.
+
+        Notifiers are transient and should not be copied.
+        """
+        id_self = id(self)
+        if id_self in memo:
+            return memo[id_self]
+
+        memo[id_self] = result = TraitListObject(
+            self.trait,
+            lambda: None,
+            self.name,
+            [copy.deepcopy(x, memo) for x in self],
         )
 
+        return result
+
     def __getstate__(self):
+        """ Get the state of the object for serialization.
+
+        Notifiers are transient and should not be serialized.
+        """
```

Comment:
Should we call `super().__getstate__` to exclude the notifiers?

Reply:
Yes